src/controllers/things_controller.js

```diff
@@ -395,74 +400,126 @@ ThingsController.ws('/:thingId/', (websocket, request) => {
       }
     });
   }
 
   function onPropertyChanged(property) {
-    if (property.device.id !== thingId) {
+    if (typeof thingId !== 'undefined' && property.device.id !== thingId) {
       return;
     }
-    sendMessage(JSON.stringify({
+    sendMessage({
+      id: property.device.id,
       messageType: Constants.PROPERTY_STATUS,
       data: {
         [property.name]: property.value,
       },
-    }));
+    });
   }
 
   function onActionStatus(action) {
-    sendMessage(JSON.stringify({
+    if (action.hasOwnProperty('thingId') &&
+        typeof thingId !== 'undefined' &&
+        action.thingId !== thingId) {
+      return;
+    }
+
+    const message = {
       messageType: Constants.ACTION_STATUS,
       data: {
         [action.name]: action.getDescription(),
       },
-    }));
+    };
+    if (action.hasOwnProperty('thingId')) {
+      message.id = action.thingId;
+    }
+    sendMessage(message);
   }
 
   function onEvent(event) {
+    if (typeof thingId !== 'undefined' && event.thingId !== thingId) {
+      return;
+    }
     if (!subscribedEventNames[event.name]) {
       return;
     }
-    sendMessage(JSON.stringify({
+    sendMessage({
+      id: event.thingId,
       messageType: Constants.EVENT,
       data: {
         [event.name]: event.getDescription(),
       },
-    }));
+    });
   }
 
-  function onConnected(connected) {
-    sendMessage(JSON.stringify({
+  function onConnected(id, connected) {
+    sendMessage({
+      id,
       messageType: Constants.CONNECTED,
       data: connected,
-    }));
+    });
   }
 
-  Things.getThing(thingId).then((thing) => {
-    thing.registerWebsocket(websocket);
+  let thingCleanups = {};
+  function addThing(thing) {
     thing.addEventSubscription(onEvent);
-    thing.addConnectedSubscription(onConnected);
+    const onConn = onConnected.bind(null, thing.id);
+    thing.addConnectedSubscription(onConn);
+    const onRem = () => {
+      if (thingCleanups[thing.id]) {
+        thingCleanups[thing.id]();
+        delete thingCleanups[thing.id];
+      }
+      if (typeof thingId !== 'undefined' &&
+          (websocket.readyState === WebSocket.OPEN ||
+           websocket.readyState === WebSocket.CONNECTING)) {
+        websocket.close();
+      }
+    };
+    thing.addRemovedSubscription(onRem);
 
-    websocket.on('close', () => {
+    const thingCleanup = () => {
       thing.removeEventSubscription(onEvent);
-      thing.removeConnectedSubscription(onConnected);
+      thing.removeConnectedSubscription(onConn);
+      thing.removeRemovedSubscription(onRem);
+    };
+    thingCleanups[thing.id] = thingCleanup;
+  }
+
+  if (typeof thingId !== 'undefined') {
+    Things.getThing(thingId).then((thing) => {
+      addThing(thing);
+    }).catch(() => {
+      console.error('WebSocket opened on nonexistent thing', thingId);
+      sendMessage({
+        messageType: Constants.ERROR,
+        data: {
+          status: '404 Not Found',
```

Comment:
A property like `code` containing only the http code would make it easier to handle errors.

Reply:
Nice idea! I've added a `code` property to all the error messages